// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: dingtalktest.proto

package pb

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on DingTalkTestRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DingTalkTestRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DingTalkTestRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DingTalkTestRequestMultiError, or nil if none found.
func (m *DingTalkTestRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *DingTalkTestRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if utf8.RuneCountInString(m.GetWebhook()) < 1 {
		err := DingTalkTestRequestValidationError{
			field:  "Webhook",
			reason: "value length must be at least 1 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if utf8.RuneCountInString(m.GetSecret()) < 1 {
		err := DingTalkTestRequestValidationError{
			field:  "Secret",
			reason: "value length must be at least 1 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return DingTalkTestRequestMultiError(errors)
	}

	return nil
}

// DingTalkTestRequestMultiError is an error wrapping multiple validation
// errors returned by DingTalkTestRequest.ValidateAll() if the designated
// constraints aren't met.
type DingTalkTestRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DingTalkTestRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DingTalkTestRequestMultiError) AllErrors() []error { return m }

// DingTalkTestRequestValidationError is the validation error returned by
// DingTalkTestRequest.Validate if the designated constraints aren't met.
type DingTalkTestRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DingTalkTestRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DingTalkTestRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DingTalkTestRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DingTalkTestRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DingTalkTestRequestValidationError) ErrorName() string {
	return "DingTalkTestRequestValidationError"
}

// Error satisfies the builtin error interface
func (e DingTalkTestRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDingTalkTestRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DingTalkTestRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DingTalkTestRequestValidationError{}

// Validate checks the field values on DingTalkTestResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DingTalkTestResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DingTalkTestResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DingTalkTestResponseMultiError, or nil if none found.
func (m *DingTalkTestResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *DingTalkTestResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Success

	// no validation rules for Error

	if len(errors) > 0 {
		return DingTalkTestResponseMultiError(errors)
	}

	return nil
}

// DingTalkTestResponseMultiError is an error wrapping multiple validation
// errors returned by DingTalkTestResponse.ValidateAll() if the designated
// constraints aren't met.
type DingTalkTestResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DingTalkTestResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DingTalkTestResponseMultiError) AllErrors() []error { return m }

// DingTalkTestResponseValidationError is the validation error returned by
// DingTalkTestResponse.Validate if the designated constraints aren't met.
type DingTalkTestResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DingTalkTestResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DingTalkTestResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DingTalkTestResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DingTalkTestResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DingTalkTestResponseValidationError) ErrorName() string {
	return "DingTalkTestResponseValidationError"
}

// Error satisfies the builtin error interface
func (e DingTalkTestResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDingTalkTestResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DingTalkTestResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DingTalkTestResponseValidationError{}
